### PR TITLE
use buildpg to prepare function

### PR DESCRIPTION
**Pull request by vincentsarago**
ref #68 

@bitner let me know if that's what you envisioned. 

~We now have another problem, we need to `cast` additional argument because at the application level we will get `string`. I'm not sure what's the best way to do this:~
- ~parse SQL code in python (seems possible using sqlparse)~
- ~use run a SQL code to get argument+type 🤷‍♂️~

### CODE REVIEW

File: tests/fixtures/squares.sql [hunk 1]
```diff
@@ -6,7 +6,7 @@ CREATE OR REPLACE FUNCTION squares(
     ymax float,
     epsg integer,
     -- additional parameters
-    depth integer default 2
+    query_params json
```
Comment (vincentsarago) [inline tests/fixtures/squares.sql:9]: this is the same approach as in https://github.com/urbica/martin#function-sources

Basically, we forward the additional parameter (`kwargs`) to the SQL function and it's up the the user to know what to do with it 🤷‍♂️

File: tests/fixtures/squares.sql [hunk 2]
```diff
@@ -26,6 +27,8 @@ BEGIN
 
     -- We want bbox divided up into depth*depth squares per bbox,
     -- so what is the width of a square?
+    depth := coalesce((query_params ->> 'depth')::int, 2);
```
Comment (vincentsarago) [inline tests/fixtures/squares.sql:30]: `example`

`depth` might not be present in the query_params, but we need a default

**Comment by vincentsarago** [comment]
Last commit changed the way we handle `function` layers. This is now similar to https://github.com/urbica/martin#function-sources, where the application doesn't know the SQL function parameters but forward additional query parameter as a JSON param. This forward the responsibility of knowing how to parse the query to the user.

IMO this is the best option we have now. As mentioned before, we don't want to `register` the function to the database nor we want to parse the SQL code using python. 

## To Do

- [ ] review from @bitner 
- [ ] update documentation

File: tests/fixtures/squares.sql (hunk 1 above)
Comment (bitner) [reply tests/fixtures/squares.sql:9]: Yes, I like this pattern (it's basically what I am doing with pgstac)

**Review by bitner — APPROVED** [review]
Looks great, definitely brings things up to snuff for best practices for sql injection.

**Pull request merged**